### ISSUE TITLE
123.2 Closes itself.

### ISSUE THREAD

**Issue by sflorg** [diffractor/diffractor#55]
Well darn Zac, this last version 123.2 is having lots of issues unfortunately.  It has a tendency to close itself a lot. Just on starting most of the time, also seems to happen if a delete images. Doesn't happen all the time, but more often than not. Starting it a few times does seem to straighten it out for awhile than right back to doing it, have anyone else been experiencing this.
Again version 123.2
OS Win 10 64 20H2 Build 19042.985

Hope all is well with you and any other info you may need I'll provide.

**Comment by Peceha** [comment]
the same happens here - mostly during startup but sometimes just after some activity from my side

**Comment by ZacWalk** [comment]
oh - sorry about that. I have stopped the auto-upgrade. I have not been able to reproduce the shutdown locally. I will keep trying though. 

If anyone has this issue can they attach a log file?  Should be in this location: C:\Users\<user>\AppData\Local\Diffractor\diffractor.log

V123.2 has a new 'location' dialog. I didn't think it a high-risk change :(

**Comment by Peceha** [comment]
Happened just now, the log file is below:
[diffractor.log](https://github.com/diffractor/diffractor/files/6525048/diffractor.log)

**Comment by ZacWalk** [comment]
thanks - "bad_weak_ptr exception" - not much to go on.

I am making a guess on what that might be. Could you try this patched version (v123.3): https://diffractor.com/diffractor-setup-test.exe

**Comment by sflorg** [comment]
Thanks Zac. Just got in and installed the test version. . . no issues yet, but will run a full stress test on it in the morning, but it is time to turn up the O2 and rest after a very long day. Sorry about the delay in responding.

**Comment by ZacWalk** [comment]
Thanks! Keep me posted. I will also add better error handling for this type of Error - just in case.

**Comment by Peceha** [comment]
doesn't close itself anymore - I've been playing with it since morning (including starting up/closing the program every now and then) - so far no problems

**Comment by sflorg** [comment]
Yep, have to concur with Peceha, the issue does appear to be resolved. Thanks Zac.

**Comment by ZacWalk** [comment]
cool - good news! 

Getting technical :

From what I can tell - in that version - the C++ code that builds the title for items section/group occasionally throws a bad_weak_ptr exception. Groups are re-built when you open a folder but also when a file is updated in the folder or the group order changes etc. 

The exception occurred when the title object referenced itself using shared_from_this. I was able to rewrite the code to remove the shared_from_this call. Seems to be fixed but still a little mysterious.

TLDR - I will add some better error handling for this :)

**Comment by sflorg** [comment]
Thanks for the explanation and glad I could help bring a little mysterious thinking into your life. 😊
Till the next time take care of yourself.
Heidi-Ann

I do have some thoughts on promoting "Diffractor" that will share with you on a later date.

**Comment by ZacWalk** [comment]
Thanks Heidi-Ann. Should not be mysterious considering I wrote the code but that is C++ :)

Sure  - let me know any promoting ideas. Diffractor is not well known - yet!

**Comment by Peceha** [comment]
I just noticed that Diffractor is in the software section on VideoHelp page - so I let myself leave a comment there :)
https://www.videohelp.com/software/Diffractor

**Comment by ZacWalk** [comment]
Cool - thanks @Peceha. videohelp is a great site.
